src/tls/test-cockpit-certificate-ensure.c

```diff
@@ -0,0 +1,432 @@
+                               SRCDIR "/test/data/expired/0-self-signed.key",
+                               NULL },
+  .check_stdout = "Found*self-signed*but it needs to be reissued*",
+  .check_exit = EXIT_FAILURE
+};
+
+static const TestCase case_mismatched = {
+  .files = (const gchar *[]) { SRCDIR "/test/data/expired/0-self-signed.cert",
+                               SRCDIR "/test/data/100years/0-self-signed.key",
+                               NULL },
+  .check_stderr = "*certificate and the given key do not match*",
+  .check_exit = EXIT_FAILURE
+};
+
+static const TestCase expired_not_selfsign = {
+  .files = (const gchar *[]) { SRCDIR "/test/data/expired/1.cert",
+                               SRCDIR "/test/data/expired/1.key",
+                               NULL },
+  .check_stdout = "Would use*1.cert*",
+  .check_exit = EXIT_SUCCESS,
+
+  .cert_source = "*/cockpit/ws-certs.d/1.cert",
+  .key_source = "*/cockpit/ws-certs.d/1.key",
+  .copy_exit = EXIT_SUCCESS,
+};
+
+static const TestCase expired_combined = {
+  .files = (const gchar *[]) { SRCDIR "/test/data/expired/combined.cert",
+                               NULL },
+  .check_stderr = "*merged certificate and key files are unsupported*",
+  .check_exit = EXIT_FAILURE,
+
+  .copy_stderr = "*merged certificate and key files are deprecated*",
+  .copy_exit = EXIT_SUCCESS,
+  .key_source = "*/cockpit/ws-certs.d/combined.cert",
+  .cert_source = "*/cockpit/ws-certs.d/combined.cert"
+};
+
+static const TestCase many_files = {
+  .files = (const gchar *[]) { SRCDIR "/test/data/expired/0-self-signed.cert",
+                               SRCDIR "/test/data/expired/0-self-signed.key",
+                               SRCDIR "/test/data/expired/1.cert",
+                               SRCDIR "/test/data/expired/1.key",
+                               SRCDIR "/test/data/expired/combined.cert",
+                               NULL },
+  .check_stderr = "*merged certificate and key files are unsupported*",
+  .check_exit = EXIT_FAILURE,
+
+  .copy_stderr = "*merged certificate and key files are deprecated*",
+  .copy_exit = EXIT_SUCCESS,
+  .key_source = "*/cockpit/ws-certs.d/combined.cert",
+  .cert_source = "*/cockpit/ws-certs.d/combined.cert"
+};
+
+int
+main (int argc,
+      char *argv[])
+{
+  cockpit_test_init (&argc, &argv);
+
+  g_test_add ("/cockpit-certificate-ensure/check/combine-good-rsa",
+              Fixture, &case_good_rsa_file, fixture_setup, test_check, fixture_teardown);
+  g_test_add ("/cockpit-certificate-ensure/copy/combine-good-rsa",
+              Fixture, &case_good_rsa_file, fixture_setup, test_copy, fixture_teardown);
+  g_test_add ("/cockpit-certificate-ensure/check/combine-good-ecc",
+              Fixture, &case_good_ecc_file, fixture_setup, test_check, fixture_teardown);
+  g_test_add ("/cockpit-certificate-ensure/copy/combine-good-ecc",
+              Fixture, &case_good_ecc_file, fixture_setup, test_copy, fixture_teardown);
```

Comment:
AFAICS these are not combined files, but have split cert/key files.

Can we please add an actual combined file as a test? We have one expired one, but no valid combined one which would be accepted by --check?

Reply:
I reuse the expired one as the combined testcase because:
 - even if i generate a valid certificate, it's only a matter of time before it's expired again anyway
 - in case the file isn't called "0-self-signed.cert" then we don't check its expiration anyway, so it doesn't matter.

With that, we do support combined files, and even have fairly robust testing around that now.

Reply:
I see, thanks! So I suppose what's left of this comment is the "combine-" test name parts, which are wrong now?

Reply:
Ya.  I removed all of those now.  Not sure why they were there in the first place.  Probably because this tested that old code path in cockpit-ws which pasted the files back together before passing them into `GTlsCertificate`?